File: go/pkg/chatgear/conn_mqtt.go
package chatgear

import (
	"context"
	"encoding/binary"
	"encoding/json"
	"fmt"
	"iter"
	"net/url"
	"strings"
	"sync"
	"time"

	"github.com/haivivi/giztoy/go/pkg/audio/codec/opus"
	"github.com/haivivi/giztoy/go/pkg/mqtt0"
)

// =============================================================================
// Wire format for stamped Opus frames
// =============================================================================
//
// StampedFrame format:
//
//	+--------+------------------+------------------+
//	| Version| Timestamp (7B)   | Opus Frame Data  |
//	| (1B)   | Big-endian ms    |                  |
//	+--------+------------------+------------------+
//
// Total header: 8 bytes

const (
	frameVersion      = 1
	stampedHeaderSize = 8
)

// stampFrame creates a stamped frame from a frame and timestamp.
func stampFrame(frame opus.Frame, t time.Time) []byte {
	stamp := t.UnixMilli()
	var buf [8]byte
	binary.BigEndian.PutUint64(buf[:], uint64(stamp))
	buf[0] = frameVersion
	return append(buf[:], frame...)
}

// unstampFrame extracts the frame and timestamp from stamped data.
// Returns ok=false if the data is invalid.
func unstampFrame(b []byte) (frame opus.Frame, t time.Time, ok bool) {
	if len(b) < stampedHeaderSize {
		return nil, time.Time{}, false
	}
	if b[0] != frameVersion {
		return nil, time.Time{}, false
	}
	var buf [8]byte
	copy(buf[1:], b[1:8])
	stamp := int64(binary.BigEndian.Uint64(buf[:]))
	t = time.UnixMilli(stamp)
	frame = opus.Frame(b[stampedHeaderSize:])
	if len(frame) < 1 {
		return nil, time.Time{}, false
	}
	return frame, t, true
}

// =============================================================================
// MQTT Client Connection
// =============================================================================

// MQTTClientConfig contains configuration for dialing an MQTT connection.
type MQTTClientConfig struct {
	// Addr is the MQTT broker address (e.g., "tcp://localhost:1883").
	Addr string

	// Scope is the topic prefix (e.g., "palr/cn").
	Scope string

	// GearID is the device identifier.
	GearID string

	// Logger is used for logging warnings and errors. If nil, DefaultLogger() is used.
	Logger Logger

	// ClientID is the MQTT client identifier. If empty, a default is generated.
	ClientID string

	// KeepAlive is the keep-alive interval in seconds. Default is 60.
	KeepAlive uint16

	// ConnectTimeout is the timeout for establishing a connection. Default is 30s.
	ConnectTimeout time.Duration
}

// DialMQTT connects to an MQTT broker and returns a client connection.
func DialMQTT(ctx context.Context, cfg MQTTClientConfig) (*MQTTClientConn, error) {
	// Normalize scope
	scope := cfg.Scope
	if scope != "" && !strings.HasSuffix(scope, "/") {
		scope += "/"
	}

	// Set defaults
	clientID := cfg.ClientID
	if clientID == "" {
		clientID = fmt.Sprintf("chatgear-%s-%d", cfg.GearID, time.Now().UnixNano()%10000)
	}
	keepAlive := cfg.KeepAlive
	if keepAlive == 0 {
		keepAlive = 60
	}
	connectTimeout := cfg.ConnectTimeout
	if connectTimeout == 0 {
		connectTimeout = 30 * time.Second
	}
	logger := cfg.Logger
	if logger == nil {
		logger = DefaultLogger()
	}

	// Parse URL to extract username/password if present
	var username string
	var password []byte
	addr := cfg.Addr
	if u, err := url.Parse(cfg.Addr); err == nil && u.User != nil {
		username = u.User.Username()
		if p, ok := u.User.Password(); ok {
			password = []byte(p)
		}
		// Reconstruct URL without userinfo for dialer
		u.User = nil
		addr = u.String()
	}

	// Connect to MQTT broker
	client, err := mqtt0.Connect(ctx, mqtt0.ClientConfig{
		Addr:           addr,
		ClientID:       clientID,
		Username:       username,
		Password:       password,
		KeepAlive:      keepAlive,
		ConnectTimeout: connectTimeout,
	})
	if err != nil {
		return nil, fmt.Errorf("chatgear/mqtt: connect: %w", err)
	}

	childCtx, cancel := context.WithCancel(ctx)
	conn := &MQTTClientConn{
		client:     client,
		ctx:        childCtx,
		cancel:     cancel,
		gearID:     cfg.GearID,
		scope:      scope,
		logger:     logger,
		opusFrames: make(chan StampedOpusFrame, 1024),
		commands:   make(chan *CommandEvent, 32),
	}

	// Subscribe to downlink topics
	audioTopic := fmt.Sprintf("%sdevice/%s/output_audio_stream", scope, cfg.GearID)
	cmdTopic := fmt.Sprintf("%sdevice/%s/command", scope, cfg.GearID)

	if err := client.Subscribe(ctx, audioTopic, cmdTopic); err != nil {
		client.Close()
		cancel()
		return nil, fmt.Errorf("chatgear/mqtt: subscribe: %w", err)
	}

	logger.InfoPrintf("subscribed to MQTT topics: audio=%s, command=%s", audioTopic, cmdTopic)

	// Start receive loop
	go conn.receiveLoop()

	return conn, nil
}

// MQTTClientConn represents a client-side connection to the server via MQTT.
// It implements both UplinkTx (send to server) and DownlinkRx (receive from server).
type MQTTClientConn struct {
	client *mqtt0.Client
	ctx    context.Context
	cancel context.CancelFunc
	gearID string
	scope  string
	logger Logger

	// Downlink channels
	opusFrames chan StampedOpusFrame
	commands   chan *CommandEvent

	mu     sync.Mutex
	closed bool
}

func (c *MQTTClientConn) receiveLoop() {
	c.logger.InfoPrintf("receiveLoop started")
	audioTopic := fmt.Sprintf("%sdevice/%s/output_audio_stream", c.scope, c.gearID)
	cmdTopic := fmt.Sprintf("%sdevice/%s/command", c.scope, c.gearID)

	for {
		select {
		case <-c.ctx.Done():
			c.logger.InfoPrintf("receiveLoop: context done")
			return
		default:
		}

		msg, err := c.client.RecvTimeout(100 * time.Millisecond)
		if err != nil {
			if c.client.IsRunning() {
				c.logger.ErrorPrintf("mqtt recv error: %v", err)
			} else {
				c.logger.InfoPrintf("receiveLoop: client stopped")
			}
			return
		}
		if msg == nil {
			continue // timeout, no message
		}

		switch msg.Topic {
		case audioTopic:
			frame, t, ok := unstampFrame(msg.Payload)
			if !ok {
				c.logger.WarnPrintf("invalid stamped frame received")
				continue
			}
			c.logger.DebugPrintf("MQTT RX audio: len=%d ts=%v", len(frame), t.Format("15:04:05.000"))
			select {
			case c.opusFrames <- StampedOpusFrame{Timestamp: t, Frame: frame}:
			default:
				// Drop frame if buffer full
			}
		case cmdTopic:
			c.logger.InfoPrintf("MQTT RX command: %s", string(msg.Payload))
			var evt CommandEvent
			if err := json.Unmarshal(msg.Payload, &evt); err != nil {
				c.logger.WarnPrintf("failed to unmarshal command: %v", err)
				continue
			}
			select {
			case c.commands <- &evt:
			default:
				c.logger.WarnPrintf("commands channel full, dropping command")
			}
		}
	}
}

// --- UplinkTx implementation ---

func (c *MQTTClientConn) SendOpusFrame(timestamp time.Time, frame opus.Frame) error {
	topic := fmt.Sprintf("%sdevice/%s/input_audio_stream", c.scope, c.gearID)
	stamped := stampFrame(frame, timestamp)
	c.logger.DebugPrintf("MQTT TX audio: len=%d ts=%v", len(frame), timestamp.Format("15:04:05.000"))
	return c.client.Publish(c.ctx, topic, stamped)
}

func (c *MQTTClientConn) SendState(state *StateEvent) error {
	topic := fmt.Sprintf("%sdevice/%s/state", c.scope, c.gearID)
	data, err := json.Marshal(state)
	if err != nil {
		return err
	}
	c.logger.InfoPrintf("MQTT TX state: %s", string(data))
	return c.client.Publish(c.ctx, topic, data)
}

func (c *MQTTClientConn) SendStats(stats *StatsEvent) error {
	topic := fmt.Sprintf("%sdevice/%s/stats", c.scope, c.gearID)
	data, err := json.Marshal(stats)
	if err != nil {
		return err
	}
	c.logger.InfoPrintf("MQTT TX stats: %s", string(data))
	return c.client.Publish(c.ctx, topic, data)
}

// --- DownlinkRx implementation ---

func (c *MQTTClientConn) OpusFrames() iter.Seq2[StampedOpusFrame, error] {
	return func(yield func(StampedOpusFrame, error) bool) {
		for {
			select {
			case <-c.ctx.Done():
				return
			case frame, ok := <-c.opusFrames:
				if !ok {
					return
				}
				if !yield(frame, nil) {
					return
				}
			}
		}
	}
}

func (c *MQTTClientConn) Commands() iter.Seq2[*CommandEvent, error] {
	return func(yield func(*CommandEvent, error) bool) {
		for {
			select {
			case <-c.ctx.Done():
				return
			case cmd, ok := <-c.commands:
				if !ok {
					return
				}
				if !yield(cmd, nil) {
					return
				}
			}
		}
	}
}

// --- Lifecycle ---

func (c *MQTTClientConn) Close() error {
	c.mu.Lock()
	if c.closed {
		c.mu.Unlock()
		return nil
	}
	c.closed = true
	c.mu.Unlock()

	c.cancel()
	return c.client.Close()
}

// GearID returns the gear ID for this connection.
func (c *MQTTClientConn) GearID() string {
	return c.gearID
}

// Compile-time interface assertions
var (
	_ UplinkTx   = (*MQTTClientConn)(nil)
	_ DownlinkRx = (*MQTTClientConn)(nil)
)

File: rust/chatgear/src/conn_mqtt.rs
//! MQTT client connection for chatgear.
//!
//! Implements `UplinkTx` (send to server) and `DownlinkRx` (receive from server)
//! over an MQTT broker, matching Go's `MQTTClientConn`.
//!
//! Wire format for stamped Opus frames:
//!
//! ```text
//! +--------+------------------+------------------+
//! | Version| Timestamp (7B)   | Opus Frame Data  |
//! | (1B)   | Big-endian ms    |                  |
//! +--------+------------------+------------------+
//! ```
//! Total header: 8 bytes

use crate::{
    ConnError, DownlinkRx, GearStateEvent, GearStatsEvent, SessionCommandEvent,
    StampedOpusFrame, UplinkTx,
    logger::{self, Logger},
};
use async_trait::async_trait;
use std::sync::Arc;
use tokio::sync::{mpsc, Mutex};
use tokio_util::sync::CancellationToken;

const FRAME_VERSION: u8 = 1;
const STAMPED_HEADER_SIZE: usize = 8;

/// Encode a StampedOpusFrame into the wire format.
pub fn stamp_frame(frame: &StampedOpusFrame) -> Vec<u8> {
    let stamp = frame.timestamp_ms as u64;
    let mut buf = stamp.to_be_bytes();
    buf[0] = FRAME_VERSION;
    let mut result = Vec::with_capacity(STAMPED_HEADER_SIZE + frame.frame.len());
    result.extend_from_slice(&buf);
    result.extend_from_slice(&frame.frame);
    result
}

/// Decode wire format into a StampedOpusFrame.
/// Returns None if the data is invalid.
pub fn unstamp_frame(data: &[u8]) -> Option<StampedOpusFrame> {
    if data.len() < STAMPED_HEADER_SIZE + 1 {
        return None;
    }
    if data[0] != FRAME_VERSION {
        return None;
    }
    let mut buf = [0u8; 8];
    buf[0] = 0; // Clear version byte
    buf[1..8].copy_from_slice(&data[1..8]);
    let stamp = i64::from_be_bytes(buf);
    let frame = data[STAMPED_HEADER_SIZE..].to_vec();
    Some(StampedOpusFrame::new(stamp, frame))
}

/// Configuration for MQTT client connection.
#[derive(Debug, Clone)]
pub struct MQTTClientConfig {
    /// MQTT broker address (e.g., "127.0.0.1:1883").
    pub addr: String,
    /// Topic prefix (e.g., "palr/cn").
    pub scope: String,
    /// Device identifier.
    pub gear_id: String,
    /// MQTT client identifier. Auto-generated if empty.
    pub client_id: String,
    /// Keep-alive interval in seconds. Default: 60.
    pub keep_alive: u16,
}

/// MQTT client connection.
///
/// Implements `UplinkTx` (send audio/state/stats to server) and
/// `DownlinkRx` (receive audio/commands from server).
pub struct MQTTClientConn {
    client: Arc<giztoy_mqtt0::Client>,
    cancel: CancellationToken,
    gear_id: String,
    scope: String,
    logger: Arc<dyn Logger>,

    opus_rx: Arc<Mutex<mpsc::Receiver<StampedOpusFrame>>>,
    commands_rx: Arc<Mutex<mpsc::Receiver<SessionCommandEvent>>>,

    closed: Arc<Mutex<bool>>,
}

impl MQTTClientConn {
    /// Returns the gear ID for this connection.
    pub fn gear_id(&self) -> &str {
        &self.gear_id
    }
}

/// Connects to an MQTT broker and returns a client connection.
pub async fn dial_mqtt(cfg: MQTTClientConfig, log: Arc<dyn Logger>) -> Result<MQTTClientConn, ConnError> {
    let scope = if cfg.scope.is_empty() || cfg.scope.ends_with('/') {
        cfg.scope.clone()
    } else {
        format!("{}/", cfg.scope)
    };

    let client_id = if cfg.client_id.is_empty() {
        format!("chatgear-{}-{}", cfg.gear_id, std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap_or_default()
            .as_millis() % 10000)
    } else {
        cfg.client_id.clone()
    };

    let keep_alive = if cfg.keep_alive == 0 { 60 } else { cfg.keep_alive };

    let mut mqtt_cfg = giztoy_mqtt0::ClientConfig::new(&cfg.addr, &client_id);
    mqtt_cfg.keep_alive = keep_alive;
    let client = giztoy_mqtt0::Client::connect(mqtt_cfg)
        .await.map_err(|e| ConnError::Other(format!("mqtt connect: {}", e)))?;

    let client = Arc::new(client);

    let audio_topic = format!("{}device/{}/output_audio_stream", scope, cfg.gear_id);
    let cmd_topic = format!("{}device/{}/command", scope, cfg.gear_id);

    client.subscribe(&[&audio_topic, &cmd_topic]).await
        .map_err(|e| ConnError::Other(format!("mqtt subscribe: {}", e)))?;

    log.info(&format!("subscribed to MQTT topics: audio={}, command={}", audio_topic, cmd_topic));

    let (opus_tx, opus_rx) = mpsc::channel(1024);
    let (cmd_tx, commands_rx) = mpsc::channel(32);
    let cancel = CancellationToken::new();

    let conn = MQTTClientConn {
        client: client.clone(),
        cancel: cancel.clone(),
        gear_id: cfg.gear_id.clone(),
        scope: scope.clone(),
        logger: log.clone(),
        opus_rx: Arc::new(Mutex::new(opus_rx)),
        commands_rx: Arc::new(Mutex::new(commands_rx)),
        closed: Arc::new(Mutex::new(false)),
    };

    // Spawn receive loop
    let recv_client = client.clone();
    let recv_cancel = cancel.clone();
    let recv_scope = scope.clone();
    let recv_gear_id = cfg.gear_id.clone();
    let recv_log = log.clone();
    tokio::spawn(async move {
        receive_loop(recv_client, recv_cancel, &recv_scope, &recv_gear_id, recv_log, opus_tx, cmd_tx).await;
    });

    Ok(conn)
}

async fn receive_loop(
    client: Arc<giztoy_mqtt0::Client>,
    cancel: CancellationToken,
    scope: &str,
    gear_id: &str,
    log: Arc<dyn Logger>,
    opus_tx: mpsc::Sender<StampedOpusFrame>,
    cmd_tx: mpsc::Sender<SessionCommandEvent>,
) {
    let audio_topic = format!("{}device/{}/output_audio_stream", scope, gear_id);
    let cmd_topic = format!("{}device/{}/command", scope, gear_id);

    loop {
        tokio::select! {
            _ = cancel.cancelled() => {
                log.info("receiveLoop: cancelled");
                return;
            }
            result = client.recv() => {
                match result {
                    Ok(msg) => {
                        if msg.topic == audio_topic {
                            if let Some(frame) = unstamp_frame(&msg.payload) {
                                let _ = opus_tx.try_send(frame);
                            } else {
                                log.warn("invalid stamped frame received");
                            }
                        } else if msg.topic == cmd_topic {
                            match serde_json::from_slice::<SessionCommandEvent>(&msg.payload) {
                                Ok(evt) => { let _ = cmd_tx.try_send(evt); }
                                Err(e) => log.warn(&format!("failed to unmarshal command: {}", e)),
                            }
                        }
                    }
                    Err(e) => {
                        log.warn(&format!("mqtt recv error: {}", e));
                        return;
                    }
                }
            }
        }
    }
}

#[async_trait]
impl UplinkTx for MQTTClientConn {
    async fn send_opus_frame(&self, frame: &StampedOpusFrame) -> Result<(), ConnError> {
        let topic = format!("{}device/{}/input_audio_stream", self.scope, self.gear_id);
        let stamped = stamp_frame(frame);
        self.client.publish(&topic, &stamped).await
            .map_err(|e| ConnError::SendFailed(e.to_string()))
    }

    async fn send_state(&self, state: &GearStateEvent) -> Result<(), ConnError> {
        let topic = format!("{}device/{}/state", self.scope, self.gear_id);
        let data = serde_json::to_vec(state)
            .map_err(|e| ConnError::SendFailed(e.to_string()))?;
        self.client.publish(&topic, &data).await
            .map_err(|e| ConnError::SendFailed(e.to_string()))
    }

    async fn send_stats(&self, stats: &GearStatsEvent) -> Result<(), ConnError> {
        let topic = format!("{}device/{}/stats", self.scope, self.gear_id);
        let data = serde_json::to_vec(stats)
            .map_err(|e| ConnError::SendFailed(e.to_string()))?;
        self.client.publish(&topic, &data).await
            .map_err(|e| ConnError::SendFailed(e.to_string()))
    }

    async fn close(&self) -> Result<(), ConnError> {
        let mut closed = self.closed.lock().await;
        if *closed {
            return Ok(());
        }
        *closed = true;
        self.cancel.cancel();
        Ok(())
    }
}

#[async_trait]
impl DownlinkRx for MQTTClientConn {
    async fn recv_opus_frame(&self) -> Result<Option<StampedOpusFrame>, ConnError> {
        let mut rx = self.opus_rx.lock().await;
        Ok(rx.recv().await)
    }

    async fn recv_command(&self) -> Result<Option<SessionCommandEvent>, ConnError> {
        let mut rx = self.commands_rx.lock().await;
        Ok(rx.recv().await)
    }

    async fn close(&self) -> Result<(), ConnError> {
        UplinkTx::close(self).await
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_stamp_unstamp_roundtrip() {
        let frame = StampedOpusFrame::new(1700000000000, vec![0xFC, 0x01, 0x02, 0x03]);
        let stamped = stamp_frame(&frame);
        assert_eq!(stamped.len(), STAMPED_HEADER_SIZE + 4);

        let decoded = unstamp_frame(&stamped).unwrap();
        assert_eq!(decoded.timestamp_ms, frame.timestamp_ms);
        assert_eq!(decoded.frame, frame.frame);
    }

    #[test]
    fn test_unstamp_invalid() {
        // Too short
        assert!(unstamp_frame(&[]).is_none());
        assert!(unstamp_frame(&[1, 2, 3]).is_none());

        // Wrong version
        let mut data = vec![0u8; 12];
        data[0] = 99;
        assert!(unstamp_frame(&data).is_none());

        // Valid header but no frame data
        let mut data = vec![0u8; STAMPED_HEADER_SIZE];
        data[0] = FRAME_VERSION;
        assert!(unstamp_frame(&data).is_none());
    }
}
